Signal/src/ViewControllers/ConversationView/Cells/OWSQuotedMessageView.m

```diff
@@ -0,0 +1,384 @@
+#import "ConversationViewItem.h"
+#import "Environment.h"
+#import "OWSBubbleStrokeView.h"
+#import "OWSMessageCell.h"
+#import "Signal-Swift.h"
+#import <SignalMessaging/OWSContactsManager.h>
+#import <SignalMessaging/SignalMessaging-Swift.h>
+#import <SignalMessaging/UIColor+OWS.h>
+#import <SignalMessaging/UIView+OWS.h>
+#import <SignalServiceKit/TSAttachmentStream.h>
+#import <SignalServiceKit/TSMessage.h>
+#import <SignalServiceKit/TSQuotedMessage.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface OWSQuotedMessageView ()
+
+@property (nonatomic, readonly) TSQuotedMessage *quotedMessage;
+@property (nonatomic, nullable, readonly) DisplayableText *displayableQuotedText;
+
+@property (nonatomic, readonly) UIFont *textMessageFont;
+
+@property (nonatomic, nullable) OWSBubbleStrokeView *boundsStrokeView;
+
+@end
+
+#pragma mark -
+
+@implementation OWSQuotedMessageView
+
++ (OWSQuotedMessageView *)quotedMessageViewForConversation:(TSQuotedMessage *)quotedMessage
+                                     displayableQuotedText:(nullable DisplayableText *)displayableQuotedText
+{
+    OWSAssert(quotedMessage);
+
+    return
+        [[OWSQuotedMessageView alloc] initWithQuotedMessage:quotedMessage displayableQuotedText:displayableQuotedText];
+}
+
++ (OWSQuotedMessageView *)quotedMessageViewForPreview:(TSQuotedMessage *)quotedMessage
+{
+    OWSAssert(quotedMessage);
+
+    DisplayableText *_Nullable displayableQuotedText = nil;
+    if (quotedMessage.body.length > 0) {
+        displayableQuotedText = [DisplayableText displayableText:quotedMessage.body];
+    }
+
+    return
+        [[OWSQuotedMessageView alloc] initWithQuotedMessage:quotedMessage displayableQuotedText:displayableQuotedText];
+}
+
+- (instancetype)initWithQuotedMessage:(TSQuotedMessage *)quotedMessage
+                displayableQuotedText:(nullable DisplayableText *)displayableQuotedText
+{
+    self = [super init];
+
+    if (!self) {
+        return self;
+    }
+
+    OWSAssert(quotedMessage);
+
+    _quotedMessage = quotedMessage;
+    _displayableQuotedText = displayableQuotedText;
+    _textMessageFont = OWSMessageCell.defaultTextMessageFont;
+
+    return self;
+}
+
+- (BOOL)hasQuotedAttachment
+{
+    return (self.quotedMessage.contentType.length > 0
+        && ![NSObject isNullableObject:self.quotedMessage.contentType equalTo:OWSMimeTypeOversizeTextMessage]);
+}
+
+- (BOOL)hasQuotedAttachmentThumbnail
+{
+    return (self.quotedMessage.contentType.length > 0
+        && ![NSObject isNullableObject:self.quotedMessage.contentType equalTo:OWSMimeTypeOversizeTextMessage] &&
+        [TSAttachmentStream hasThumbnailForMimeType:self.quotedMessage.contentType]);
+}
+
+- (NSString *)quotedSnippet
+{
+    if (self.displayableQuotedText.displayText.length > 0) {
+        return self.displayableQuotedText.displayText;
+    } else {
+        NSString *mimeType = self.quotedMessage.contentType;
+
+        if (mimeType.length > 0) {
+            return [TSAttachment emojiForMimeType:mimeType];
```

Comment:
Per the spec, ultimately we want to use the word "Image", "GIF", "Video", right?

Reply:
It's in the next PR.